# Решение систем

In [1]:
from sympy import symbols, solve, Eq, init_printing

init_printing()  # Для красивого вывода

def solve_system(equations, variables):
    """
    Решает систему уравнений и выводит результат
    """
    try:
        solutions = solve(equations, variables)
        
        if not solutions:
            print("Система не имеет решений")
        elif isinstance(solutions, dict):
            print("Единственное решение:")
            for var, value in solutions.items():
                print(f"{var} = {value}")
        else:
            print(f"Найдено {len(solutions)} решений:")
            for i, sol in enumerate(solutions, 1):
                print(f"Решение {i}:")
                for var, value in zip(variables, sol):
                    print(f"  {var} = {value}")
                
    except Exception as e:
        print(f"Ошибка при решении: {e}")

# Пример использования
x, y, z = symbols('x y z')

# Система уравнений
equations = [
    Eq(2*x + 3*y - z, 1),
    Eq(x - y + z**3, 8),
    Eq(3*x + 2*y + z, 9)
]

variables = (x, y, z)

solve_system(equations, variables)

Найдено 3 решений:
Решение 1:
  x = 5
  y = -3
  z = 0
Решение 2:
  x = 5 - sqrt(2)
  y = -3 + sqrt(2)
  z = sqrt(2)
Решение 3:
  x = sqrt(2) + 5
  y = -3 - sqrt(2)
  z = -sqrt(2)


# Разложение в ряд Тейлора

In [ ]:
from sympy import symbols, series, sin, cos, exp, pi, pprint, latex

x = symbols('x')

# Разложение нескольких функций
functions = [
    #sin(x),
   # cos(x), 
    #exp(x),
    #exp(-x**2),
    sin(x)*cos(x) + 2*x**2
]

print("Разложения в ряд Тейлора:")
for i, func in enumerate(functions, 1):
    print(f"\n{i}. f(x) = {func}")
    expansion = series(func, x, 0, 8)
    pprint(expansion)  # Красивое отображение
    print("LaTeX код:")
    print(f"{func} = {latex(expansion)}")
    print()

Разложения в ряд Тейлора:

1. f(x) = 2*x**2 + sin(x)*cos(x)
              3      5      7        
       2   2⋅x    2⋅x    4⋅x     ⎛ 8⎞
x + 2⋅x  - ──── + ──── - ──── + O⎝x ⎠
            3      15    315         
LaTeX код:
              3      5      7        
       2   2⋅x    2⋅x    4⋅x     ⎛ 8⎞
x + 2⋅x  - ──── + ──── - ──── + O⎝x ⎠
            3      15    315         
2*x**2 + sin(x)*cos(x) = \text{None}



## Красивый латех

\begin{equation}
2*x**2 + sin(x)*cos(x) = x + 2 x^{2} - \frac{2 x^{3}}{3} + \frac{2 x^{5}}{15} - \frac{4 x^{7}}{315} + O\left(x^{8}\right)

\end{equation}

# Приведение подобных

In [14]:
from sympy import symbols, collect, factor, pprint

def simplify_expression(expr, variables=None):
    """
    Умное упрощение выражения с приведением подобных
    """
    print("Исходное выражение:")
    pprint(expr)
    
    # Автоматический выбор метода
    if variables:
        # Приведение по указанным переменным
        result = collect(expr, variables)
        print(f"\nПосле приведения по {variables}:")
    else:
        # Полная факторизация
        result = factor(expr)
        print("\nПосле полной факторизации:")
    
    pprint(result)
    return result

# Пример использования
x, y, z = symbols('x y z')
expr = 4*x**2*y + 2*x*y**2 + 6*x**2*z + 3*x*y*z + 8*x*y + 4*x*z

# Автоматическое упрощение
simplified = simplify_expression(expr, [x, y])

Исходное выражение:
   2        2          2                          
4⋅x ⋅y + 6⋅x ⋅z + 2⋅x⋅y  + 3⋅x⋅y⋅z + 8⋅x⋅y + 4⋅x⋅z

После приведения по [x, y]:
 2                 ⎛   2                    ⎞
x ⋅(4⋅y + 6⋅z) + x⋅⎝2⋅y  + 3⋅y⋅z + 8⋅y + 4⋅z⎠


In [15]:
from sympy import symbols, together, apart, collect

x = symbols('x')

# Дробное выражение
expr = (3*x + 2)/(x + 1) + (5*x - 3)/(x + 1) - (2*x + 1)/(x + 1)
print("Исходная дробь:")
print(expr)

# Приведение к общему знаменателю
combined = together(expr)
print(f"\nПосле приведения: {combined}")

# Разложение на простые дроби
decomposed = apart(combined)
print(f"Разложение: {decomposed}")

# Упрощение числителя
simplified = collect(combined, x)
print(f"Упрощенное: {simplified}")

Исходная дробь:
-(2*x + 1)/(x + 1) + (3*x + 2)/(x + 1) + (5*x - 3)/(x + 1)

После приведения: 2*(3*x - 1)/(x + 1)
Разложение: 6 - 8/(x + 1)
Упрощенное: 2*(3*x - 1)/(x + 1)


In [16]:
from sympy import symbols, Eq, collect, solve, pprint

x, y = symbols('x y')

# Система уравнений с подобными слагаемыми
eq1 = Eq(3*x + 2*y - 5*x + 7*y, 15)
eq2 = Eq(2*x - 3*y + 4*x + y, 10)

print("Исходная система:")
pprint(eq1)
pprint(eq2)

# Приведение подобных в каждом уравнении
eq1_simple = Eq(collect(eq1.lhs, [x, y]), eq1.rhs)
eq2_simple = Eq(collect(eq2.lhs, [x, y]), eq2.rhs)

print("\nПосле приведения подобных:")
pprint(eq1_simple)
pprint(eq2_simple)

# Решение упрощенной системы
solution = solve((eq1_simple, eq2_simple), (x, y))
print(f"\nРешение: {solution}")

Исходная система:
-2⋅x + 9⋅y = 15
6⋅x - 2⋅y = 10

После приведения подобных:
-2⋅x + 9⋅y = 15
6⋅x - 2⋅y = 10

Решение: {x: 12/5, y: 11/5}
